libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h

```diff
@@ -0,0 +1,87 @@
+            AP_HAL::panic("Can only be one AC_AttitudeControl_Multi_6DoF");
+        }
+        _singleton = this;
+    }
+
+    static AC_AttitudeControl_Multi_6DoF *get_singleton() {
+        return _singleton;
+    }
+
+    /*
+        override input functions to attitude controller and convert desired angles into thrust angles and substitute for osset angles
+    */
+
+    // Command an euler roll and pitch angle and an euler yaw rate with angular velocity feedforward and smoothing
+    void input_euler_angle_roll_pitch_euler_rate_yaw(float euler_roll_angle_cd, float euler_pitch_angle_cd, float euler_yaw_rate_cds)  override ;
+
+    // Command an euler roll, pitch and yaw angle with angular velocity feedforward and smoothing
+    void input_euler_angle_roll_pitch_yaw(float euler_roll_angle_cd, float euler_pitch_angle_cd, float euler_yaw_angle_cd, bool slew_yaw) override;
+
+    /*
+        all other input functions should zero thrust vectoring and behave as a normal copter
+    */
+
+    // Command euler yaw rate and pitch angle with roll angle specified in body frame
+    // (used only by tailsitter quadplanes)
+    void input_euler_rate_yaw_euler_angle_pitch_bf_roll(bool plane_controls, float euler_roll_angle_cd, float euler_pitch_angle_cd, float euler_yaw_rate_cds) override;
+
+    // Command an euler roll, pitch, and yaw rate with angular velocity feedforward and smoothing
+    void input_euler_rate_roll_pitch_yaw(float euler_roll_rate_cds, float euler_pitch_rate_cds, float euler_yaw_rate_cds) override;
+
+     // Command an angular velocity with angular velocity feedforward and smoothing
+    void input_rate_bf_roll_pitch_yaw(float roll_rate_bf_cds, float pitch_rate_bf_cds, float yaw_rate_bf_cds) override;
+
+    // Command an angular velocity with angular velocity feedforward and smoothing
+    void input_rate_bf_roll_pitch_yaw_2(float roll_rate_bf_cds, float pitch_rate_bf_cds, float yaw_rate_bf_cds) override;
+
+    // Command an angular velocity with angular velocity smoothing using rate loops only with integrated rate error stabilization
+    void input_rate_bf_roll_pitch_yaw_3(float roll_rate_bf_cds, float pitch_rate_bf_cds, float yaw_rate_bf_cds) override;
+
+    // Command an angular step (i.e change) in body frame angle
+    void input_angle_step_bf_roll_pitch_yaw(float roll_angle_step_bf_cd, float pitch_angle_step_bf_cd, float yaw_angle_step_bf_cd) override;
+
+    // run lowest level body-frame rate controller and send outputs to the motors
+    void rate_controller_run() override;
+
+    // limiting lean angle based on throttle makes no sense for 6DoF, always allow 90 deg, return in centi-degrees
+    float get_althold_lean_angle_max() const override { return 9000.0f; }
+
+    // set the attitude that will be used in 6DoF flight
+    void set_offset_roll_pitch(float roll_deg, float pitch_deg) {
+        roll_offset_deg = roll_deg;
+        pitch_offset_deg = pitch_deg;
+    }
+
+    void set_forward_enable(bool b) {
```

Comment:
It would be good to have a comment above "set_forward_enable()" and "set_lateral_enable()"